orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1192 @@
+
+    /*
+     * Iterate through the given group members add them to be removed.
+     *
+     * Keep track of the SAI remove statuses in case one of them returns an
+     * error.  We assume that removal should always succeed.  If for some
+     * reason it doesn't, there's nothing we can do, but we'll log an error
+     * later.
+     */
+    std::map<NextHopKey, sai_status_t> statuses;
+
+    for (const auto& nh_key : nh_keys)
+    {
+        const NextHopGroupMember& nhgm = m_members.at(nh_key);
+
+        if (nhgm.isSynced())
+        {
+            nextHopGroupMemberBulker.remove_entry(&statuses[nh_key],
+                                                    nhgm.getGmId());
+        }
+    }
+
+    /* Flush the bulker to apply the changes. */
+    nextHopGroupMemberBulker.flush();
+
+    /*
+     * Iterate over the statuses map and check if the removal was successful.
+     * If it was, decrement the Crm counter and reset the member's ID.  If it
+     * wasn't, log an error message.
+     */
+    bool success = true;
+
+    for (const auto& status : statuses)
+    {
+        if (status.second == SAI_STATUS_SUCCESS)
+        {
+            m_members.at(status.first).remove();
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Could not remove next hop group member %s, rv: %d",
+                            status.first.to_string().c_str(), status.second);
+            success = false;
+        }
+    }
+
+    return success;
+}
+
+/*
+ * Purpose:     Update the next hop group based on a new next hop group key.
+ * Description: Update the group's members by removing the members that aren't
+ *              in the new next hop group and adding the new members.  We first
+ *              remove the missing members to avoid cases where we reached the
+ *              ASIC group members limit.  This will not update the group's SAI
+ *              ID in any way, unless we are promoting a temporary group.
+ * Params:      IN  nhg_key - The new next hop group key to update to.
+ * Returns:     true, if the operation was successful;
+ *              false, otherwise.
+ */
+bool NextHopGroup::update(const NextHopGroupKey& nhg_key)
+{
+    SWSS_LOG_ENTER();
+
+    /*
+     * There are three cases where the SAI ID of the NHG will change:
+     *  - changing a single next hop group to another single next hop group
+     *  - changing a single next hop group to a multiple next hop group
+     *  - changing a multiple next hop group to a single next hop group
+     *
+     * For these kind of updates, we can simply swap the existing group with
+     * the updated group, as we have no way of preserving the existing SAI ID.
+     *
+     * Also, we can perform the same operation if the group is not synced at
+     * all.
+     */
+    if ((nhg_key.getSize() == 1) || (m_members.size() == 1) || !isSynced())
+    {
+        bool was_synced = isSynced();
+        *this = NextHopGroup(nhg_key);
```

Comment:
this is bit cryptic. Can you explicitly mention the destructor will remove the older next hop object.

Reply:
Done.